### PR TITLE
Refactored ResolveAliases transformer.

### PR DESCRIPTION
**Pull request by azreika**
* Moved the transformer out of `AstTransforms.cpp` into `ResolveAliasesTransformer.cpp`
* Cleaned up the codebase
* Made the transformer more consistent with other transformers

### CODE REVIEW

**Comment by azreika** [comment]
Tests failing because of `semantic/plan2`. The semantic check on valid plan versioning seems to not be working now, even though alias resolution doesn't affect this program.

File: src/ResolveAliasesTransformer.cpp [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+    // II) compute unifying substitution
+    Substitution substitution;
+
+    // a utility for processing newly identified mappings
+    auto newMapping = [&](const std::string& var, const AstArgument* term) {
+        // found a new substitution
+        Substitution newMapping(var, term);
+
+        // apply substitution to all remaining equations
+        for (auto& eqn : equations) {
```
Comment (mmcgr) [inline src/ResolveAliasesTransformer.cpp]: eqn -> equation

File: src/ResolveAliasesTransformer.cpp [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+            eqn.apply(newMapping);
+        }
+
+        // add mapping v -> t to substitution
+        substitution.append(newMapping);
+    };
+
+    while (!equations.empty()) {
+        // get next equation to compute
+        Equation eqn = equations.back();
```
Comment (mmcgr) [inline src/ResolveAliasesTransformer.cpp]: eqn -> equation

File: src/ResolveAliasesTransformer.cpp [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+    // get list of atoms
+    std::vector<AstAtom*> atoms = res->getAtoms();
+
+    // find all functors in atoms
+    std::vector<const AstArgument*> terms;
+    for (const AstAtom* atom : atoms) {
+        for (const AstArgument* arg : atom->getArguments()) {
+            // ignore if not a functor
+            if (!dynamic_cast<const AstFunctor*>(arg)) {
```
Comment (mmcgr) [inline src/ResolveAliasesTransformer.cpp]: == nullptr

File: src/ResolveAliasesTransformer.cpp [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+                continue;
+            }
+
+            // add this functor if not seen yet
+            if (!any_of(terms, [&](const AstArgument* curr) { return *curr == *arg; })) {
```
Comment (mmcgr) [inline src/ResolveAliasesTransformer.cpp]: `cur` - we don't need two abbreviations of current

File: src/ResolveAliasesTransformer.cpp [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+ * A substitution maps variables to terms and can be applied as a transformation
+ * to AstArguments.
+ */
+class Substitution {
+    // map type used for internally storing var->term mappings
+    //      - note: variables are identified by their names
+    using map_t = std::map<std::string, std::unique_ptr<AstArgument>>;
+
+    // the mapping of variables to terms
+    map_t map;
```
Comment (mmcgr) [inline src/ResolveAliasesTransformer.cpp]: Could this be better named with nameToArgument or something similar?

File: src/ResolveAliasesTransformer.cpp [hunk 6]
```diff
@@ -0,0 +1,452 @@
+/*
+ * Souffle - A Datalog Compiler
+ * Copyright (c) 2018, The Souffle Developers. All rights reserved.
```
[the remark this answers is not in the document]
Comment (olligobber) [reply src/ResolveAliasesTransformer.cpp]: The commit that added this file (e17dfba227f71d42d33c7258558774d310a2ea35) was in 2018.

[the remark this answers is not in the document]
Comment (mmcgr) [reply src/ResolveAliasesTransformer.cpp]: Sounds good to me then

**Pull request merged**